users/handler.go

```diff
@@ -0,0 +1,172 @@
+package users
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"reflect"
+	"strconv"
+	"time"
+
+	"github.com/imadu/e-commerce-api/db"
+	"github.com/imadu/e-commerce-api/util"
+	"github.com/labstack/echo"
+	"github.com/labstack/gommon/log"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"golang.org/x/crypto/bcrypt"
+)
+
+var dbName = os.Getenv("DB_NAME")
+
+var userCollection = db.Client.Database(dbName).Collection("users")
+
+func hashPassword(password string) (string, error) {
+	bytes, err := bcrypt.GenerateFromPassword([]byte(password), 10)
+	return string(bytes), err
+}
+
+func checkPasswordHash(password, hash string) bool {
+	err := bcrypt.CompareHashAndPassword([]byte(hash), []byte(password))
+	return err == nil
+}
+
+func getUser(username string) User {
+	var result User
+	filter := bson.D{primitive.E{Key: "username", Value: username}}
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	err := userCollection.FindOne(ctx, filter).Decode(&result)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return result
+}
+
+//GetUser returns the user by id
+func GetUser(c echo.Context) error {
+	var result User
+	id := c.Param("id")
+	filter := bson.D{primitive.E{Key: "_id", Value: id}}
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	err := userCollection.FindOne(ctx, filter).Decode(&result)
+	if err != nil {
+		c.Response().WriteHeader(http.StatusNotFound)
+		return util.SendError(c, "404", "User does not exist", "failed")
+	}
+
+	return util.SendData(c, result)
+}
+
+// GetUsers returns all the users
+func GetUsers(c echo.Context) error {
+	p := c.QueryParam("page")
+	l := c.QueryParam("limit")
+
+	page, _ := strconv.ParseInt(p, 10, 64)
+
+	limit, _ := strconv.ParseInt(l, 10, 64)
+
+	findOptions := options.Find()
+	findOptions.SetLimit(limit)
+	findOptions.SetSkip(page * limit)
+
+	var results []*User
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	cur, err := userCollection.Find(ctx, bson.D{{}}, findOptions)
+	if err != nil {
+		c.Response().WriteHeader(http.StatusNotFound)
+		return util.SendError(c, "500", "could not find users", "failed")
+	}
+
+	for cur.Next(ctx) {
+		var user User
+		err := cur.Decode(&user)
+		if err != nil {
+			log.Errorf("something went wrong", err)
+		}
+		results = append(results, &user)
+	}
+
+	if err := cur.Err(); err != nil {
+		log.Fatal(err)
+	}
+	cur.Close(ctx)
+	return util.SendData(c, results)
+
+}
+
+// CreateUser creates a new user
+func CreateUser(c echo.Context) error {
```

Comment:
Same comment as to when you are creating a product, and it seems some repetitions can be abstracted